diff -r 0b4f135cdbe8 indra/newview/llvoavatar.cpp
--- a/indra/newview/llvoavatar.cpp	Thu Apr 04 09:18:46 2013 -0700
+++ b/indra/newview/llvoavatar.cpp	Sat Apr 06 05:18:35 2013 -0700
@@ -4336,6 +4336,7 @@
 		gGL.setAlphaRejectSettings(LLRender::CF_DEFAULT);
 	}
 
+#if !LL_DARWIN
 	if (!isSelf() || gAgent.needsRenderHead() || LLPipeline::sShadowRender)
 	{
 		if (LLPipeline::sImpostorRender)
@@ -4362,7 +4363,8 @@
 			gGL.setAlphaRejectSettings(LLRender::CF_DEFAULT);
 		}
 	}
-
+#endif
+	
 	return num_indices;
 }
 
